CONTRIBUTING.md

```diff
@@ -27,8 +27,90 @@ Realm welcomes all contributions! The only requirement we have is that, like man
 
 ## Unit tests
 
-All PR's must be accompanied by related unit tests.
-
+All PR's must be accompanied by related unit tests. All bug fixes must have a unit test proving that the bug is fixed.
+You can use `./realm/gradlew connectedCheck createDebugCoverageReport` to generate a coverage report to check for 
+missing unit test coverage. The aim is 100% code coverage.
+
+When writing unit tests, use the following guide lines:
+
+1) Unit tests must be written using JUnit4.
+
+2) All tests for a class should be grouped in a class called `<className>Tests`, unless the functionality is cross-
+   cutting like [`RxJavaTests`](https://github.com/realm/realm-java/blob/master/realm/realm-library/src/androidTest/java/io/realm/RxJavaTests.java) 
+   or [`RealmAsyncQueryTests`](https://github.com/realm/realm-java/blob/master/realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java).
+
+3) Test methods should use camelCase and underscore `_` between logical sections to increase method name readability. 
+   Methods should ideally start with the name of the method being tested. Patterns like: `<methodName>_<description>`, 
```

Comment:
Sometimes you are not just testing one method, but a sequence of events, in which case starting with a methodName would just be weird.